macrosynergy/signal/signal_return.py

```diff
@@ -33,46 +34,117 @@ class SignalReturnRelations:
         This must be one of 'D', 'W', 'M', 'Q', 'A'. Default is 'M'.
         The return series will always be summed over the sample period.
         The signal series will be aggregated according to the value of agg_sig.
-    :param <str> agg_sig: aggregation method applied to the signal value in down-
+    :param <str> agg_sig: aggregation method applied to the signal values in down-
         sampling. The default is "last".
+        If defined, the additional signals will also use the same aggregation method for
+        any down-sampling.
     :param <int> fwin: forward window of return category in base periods. Default is 1.
         This conceptually corresponds to the holding period of a position in
         accordance with the signal.
 
+    NB:.
+    The Class can also facilitate the analysis of additional signals which are passed
+    into the parameter 'rival_sigs'.
     """
-    def __init__(self, df: pd.DataFrame, ret: str, sig: str, cids: List[str] = None,
-                 sig_neg: bool = False, start: str = None, end: str = None,
-                 fwin: int = 1, blacklist: dict = None, agg_sig: str = 'last',
-                 freq: str = 'M'):
+    def __init__(self, df: pd.DataFrame, ret: str, sig: str, rival_sigs: List[str] = None,
+                 cids: List[str] = None, sig_neg: bool = False, start: str = None,
+                 end: str = None, fwin: int = 1, blacklist: dict = None,
+                 agg_sig: str = 'last', freq: str = 'M'):
 
         self.dic_freq = {'D': 'daily', 'W': 'weekly', 'M': 'monthly',
                          'Q': 'quarterly', 'A': 'annual'}
         self.metrics = ['accuracy', 'bal_accuracy', 'pos_sigr', "pos_retr",
                         'pos_prec', 'neg_prec', 'pearson', 'pearson_pval',
                         'kendall', 'kendall_pval']
 
+        self.ret = ret
+        self.freq = freq
+
+        self.start = start
+        self.end = end
+        self.blacklist = blacklist
+        self.fwin = fwin
+
+        # Use the sum as the default aggregation method to show the true return over the
+        # time-period.
         self.df = categories_df(df, xcats=[sig, ret], cids=cids, val='value',
                                 start=start, end=end, freq=freq, blacklist=blacklist,
                                 lag=1, fwin=fwin, xcat_aggs=[agg_sig, 'sum'])
 
+        self.cids = list(np.sort(self.df.index.get_level_values(0).unique()))
+
+        if rival_sigs is not None:
+
+            if isinstance(rival_sigs, str):
+                rival_sigs = [rival_sigs]
+            assert isinstance(rival_sigs, list), "List of signal(s) expected."
+            xcats = list(df['xcat'].unique())
+
+            df_sigs = self.__signal_table__(df=dfd, sigs=rival_sigs, xcats=xcats,
+                                            agg_sig=agg_sig)
+            # Todo: dfd has nor reference, should this be df?
```

Comment:
Typo - should be "df". Problem with Python that it will still reference the variable from the parent scope, as opposed to the locally passed DataFrame.